discovery/store.go

```diff
@@ -0,0 +1,379 @@
+	CredentialType *string
+	Properties     []credentialPropertyRecord `gorm:"foreignKey:CredentialID;references:ID"`
+}
+
+// TableName returns the table name for this DTO.
+func (p credentialRecord) TableName() string {
+	return "discovery_credential"
+}
+
+// credentialPropertyRecord is a property of a Verifiable Credential in a Verifiable Presentation in a discovery service.
+type credentialPropertyRecord struct {
+	// CredentialID refers to the entry record in discovery_credential
+	CredentialID string `gorm:"primaryKey"`
+	// Key is JSON path of the property.
+	Key string `gorm:"primaryKey"`
+	// Value is the value of the property.
+	Value string
+}
+
+// TableName returns the table name for this DTO.
+func (l credentialPropertyRecord) TableName() string {
+	return "discovery_credential_prop"
+}
+
+type sqlStore struct {
+	db        *gorm.DB
+	writeLock sync.Mutex
+}
+
+func newSQLStore(db *gorm.DB, definitions map[string]Definition) (*sqlStore, error) {
+	// Creates entries in the discovery service table with initial timestamp, if they don't exist yet
+	for _, definition := range definitions {
+		currentList := serviceRecord{
+			ID: definition.ID,
+		}
+		if err := db.FirstOrCreate(&currentList, "id = ?", definition.ID).Error; err != nil {
+			return nil, err
+		}
+	}
+	return &sqlStore{
+		db:        db,
+		writeLock: sync.Mutex{},
+	}, nil
+}
+
+// Add adds a presentation to the list of presentations.
+// Timestamp should be passed if the presentation was received from a remote Discovery Server, then it is stored alongside the presentation.
+// If the local node is the Discovery Server and thus is responsible for the timestamping,
+// nil should be passed to let the store determine the right value.
+func (s *sqlStore) add(serviceID string, presentation vc.VerifiablePresentation, timestamp *Timestamp) error {
+	credentialSubjectID, err := credential.PresentationSigner(presentation)
+	if err != nil {
+		return err
+	}
+	if exists, err := s.exists(serviceID, credentialSubjectID.String(), presentation.ID.String()); err != nil {
+		return err
+	} else if exists {
+		return ErrPresentationAlreadyExists
+	}
+	if _, isSQLite := s.db.Config.Dialector.(*sqlite.Dialector); isSQLite {
+		// SQLite does not support SELECT FOR UPDATE and allows only 1 active write transaction at any time,
+		// and any other attempt to acquire a write transaction will directly return an error.
+		// This is in contrast to most other SQL-databases, which let the 2nd thread wait for some time to acquire the lock.
+		// The general advice for SQLite is to retry the operation, which is just poor-man's scheduling.
+		// So to keep behavior consistent across databases, we'll just lock the entire store for the duration of the transaction.
+		// See https://github.com/nuts-foundation/nuts-node/pull/2589#discussion_r1399130608
+		s.writeLock.Lock()
+		defer s.writeLock.Unlock()
+	}
+	if err := s.prune(); err != nil {
+		return err
+	}
```

Comment:
I thought I could spare the complexity, and as long as the column is indexed it should be really, really fast.

Reply:
it depends how care orgs will be allowed to access discovery services. As long as care orgs can't use use-cases that aren't configured in a service definition I suppose there is no issue.

rationale: If a node only acts as client (applies to most) and all service definitions are gone after a restart (config error), the db will only contain outdated info. The client will also no longer publish its clients to the discovery services. This could result in weird situations where orgs can find orgs managed by other vendors but not their own, etc..

Reply:
If the service definition is wrong, which causes an error on startup. And clients should only be able to search loaded services (validation check in still-missing Search() function in module.go).
If it can't post presentations to the server, healthchecks/logging should show this error.